processing: Move to gradle-based build

neobrain opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This drops use of upstream's deprecated ant build system in favor of the Gradle build scripts [introduced in Processing 4.4](https://github.com/processing/processing4/releases/tag/processing-1300-4.4.0). 

Doing so has the nice side effect of enabling some [previously broken functionality](https://github.com/NixOS/nixpkgs/issues/441815#issuecomment-3281628746) such as examples not being bundled.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - ~~[ ] [NixOS tests] in [nixos/tests].~~
  - ~~[ ] [Package tests] at `passthru.tests`.~~
  - ~~[ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.~~
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - ~~[ ] Package update: when the change is major or breaking.~~
- NixOS Release Notes
  - ~~[ ] Module addition: when adding a new NixOS module.~~
  - ~~[ ] Module update: when the change is significant.~~
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

neobrain committed:
processing: move to gradle-based build

neobrain commented on pkgs/by-name/pr/processing/package.nix:
```diff
@@ -107,38 +137,36 @@ stdenv.mkDerivation rec {
     ln -sf ${batik}/* java/libraries/svg/library/
     cp java/libraries/svg/library/share/java/batik-all-${batik.version}.jar java/libraries/svg/library/batik.jar
-    echo "tarring ffmpeg"
-    tar --checkpoint=10000 -czf build/shared/tools/MovieMaker/ffmpeg-5.0.1.gz ${ffmpeg}
```
Note: I wasn't sure why this was needed before. Is it still required?

fxzzi commented:
> Upstream generates a binary called `Processing` instead of `processing` now, which breaks `nix run nixpkgs#processing` (since that case-sensitively looks for `bin/processing`). How does one typically resolve this? Is there an attribute I can override, or should we just rename the binary back to `processing` (overriding upstream)?

set meta.mainProgram to `Processing`

neobrain committed:
processing: set meta.mainProgram

neobrain commented on pkgs/by-name/pr/processing/skip-distributable.patch:
```diff
@@ -0,0 +1,24 @@
+diff --git a/app/build.gradle.kts b/app/build.gradle.kts
```
@Stefterv Without this patch, the build fails since it somehow doesn't find the createDistributable task. Does this make any sense to you?

```
processing> Running phase: buildPhase
processing> +++ gradle createDistributable
...
processing> FAILURE: Build failed with an exception.
processing> 
processing> * Where:
processing> Build file '/build/source/app/build.gradle.kts' line: 536
processing> 
processing> * What went wrong:
processing> Could not create task ':app:zipDistributable'.
processing> > Task with name 'createDistributable' not found in project ':app'.
```

We don't need the zip distributable either way, but its built automatically as part of "gradle assemble/createDistributable". Here are some options:
* Figure out where that error comes from and fix it (and then build the zip distributable and throw it away)
* Use another command than `gradle assemble/createDistributable` that only builds required files (if that exists?)
* Keep this ugly patch in nixpkgs to skip the target

neobrain replied:
Just cleaned up the existing approach and added a comment about it for now.

neobrain committed:
processing: pin version of processing-website and processing-example repos

Stefterv commented on pkgs/by-name/pr/processing/package.nix:
```diff
@@ -107,38 +137,36 @@ stdenv.mkDerivation rec {
+    cp -dpr app/build/compose/binaries/main/app/Processing/bin $out/unwrapped
+
     mkdir -p $out/share/applications/
     cp -dp build/linux/${pname}.desktop $out/share/applications/
```
this .desktop file is old and will be deleted soon

neobrain committed:
processing: symlink uppercase executable name to lowercase

Upstream changed the executable name to uppercase. Creating a symlink
with the old name prevents breaking existing scripts.

evan-goode reviewed (approved):
LGTM, thank you!

neobrain commented:
@uninsane Apologies for the ping, but any chance this could be queued for merging? The change enables future package updates (while also fixing functionality that was broken before) and has approval by upstream & by the nixpkgs maintainer.

uninsane commented on pkgs/by-name/pr/processing/package.nix:
i see. so the process to update the package as a whole is: bump `version` attribute & update `src.hash`, then manually update the refs here by fetching tip on `processing-examples` + `processing-website` repos, then invoke the `mitmCache`'s updateScript.

neobrain commented:
Great, thanks again for the reviews everyone :)
